PopState event should be fired synchronously, even before the load event

cdumez opened the pull request:
#### 9497f1badb3df074553143a369942473cf13c10d
<pre>
PopState event should be fired synchronously, even before the load event
<a href="https://bugs.webkit.org/show_bug.cgi?id=245153">https://bugs.webkit.org/show_bug.cgi?id=245153</a>

Reviewed by Brent Fulgham.

PopState event should be fired synchronously, even before the load event:
- <a href="https://github.com/whatwg/html/issues/1792">https://github.com/whatwg/html/issues/1792</a>

We used to delay PopState events until the load event has fired but this
doesn&apos;t match other Blink or Gecko.

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/event-order/before-load-hash-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/event-order/before-load-hash-twice-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/event-order/pushState-inside-popstate-expected.txt:
Rebaseline WPT tests that are now passing or failing a little further. The ones that are still failing and due to the fact that
we fire the load event synchronously instead of queuing a task. As a result, it may fire before hashchange events that were
scheduled before the load has completed. I plan to look into this in a follow-up.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::implicitClose):
(WebCore::Document::statePopped):
* Source/WebCore/dom/Document.h:

Canonical link: <a href="https://commits.webkit.org/254519@main">https://commits.webkit.org/254519@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41cce0ab1079ba24c952798a4221d51f8a6ce63a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89204 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33765 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20021 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98520 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154833 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32271 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27817 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92989 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94851 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25632 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76125 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25562 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80497 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80470 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68536 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30049 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14533 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29773 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15496 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3174 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33220 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38455 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31923 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34587 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-history-interface/history_forward_cross_realm_method-expected.txt:
```diff
@@ -1,4 +1,6 @@
 
 
+Harness Error (TIMEOUT), message = null
```
It's always troubling to land an expected TIMEOUT. I worry we are further slowing our already glacial test system.

brentfulgham reviewed (approved):
It's great to see more progressions on WPT, and this doesn't seem to regress any other behavior. r=me.

cdumez committed:
PopState event should be fired synchronously, even before the load event
https://bugs.webkit.org/show_bug.cgi?id=245153

Reviewed by Brent Fulgham.

PopState event should be fired synchronously, even before the load event:
- https://github.com/whatwg/html/issues/1792

We used to delay PopState events until the load event has fired but this
doesn't match other Blink or Gecko.

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/event-order/before-load-hash-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/event-order/before-load-hash-twice-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/history-traversal/event-order/pushState-inside-popstate-expected.txt:
Rebaseline WPT tests that are now passing or failing a little further. The ones that are still failing and due to the fact that
we fire the load event synchronously instead of queuing a task. As a result, it may fire before hashchange events that were
scheduled before the load has completed. I plan to look into this in a follow-up.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::implicitClose):
(WebCore::Document::statePopped):
* Source/WebCore/dom/Document.h:

Canonical link: https://commits.webkit.org/254519@main
